### PR TITLE
♻️ Full migration from app router PLUS next15/react19

### PR DESCRIPTION
**Pull request by KemingHe**
Completes #375

## Part 1/2 Page Router -> App Router

- [x] All existing Playwright behavior testing passing. 5b60e5b
- [x] Added new Playwright screenshot tests to all 4 image endpoints. 74809d6
- [x] Updated unit tests to reflect the new architectural changes.

## Part 2/2 -> Next15/React19 via Official Codemod

- [x] Breezed through Codemod, fully upgraded deps and types, no major/minor code changes.
- [x] Updated typing based on latest docs.
- [x] Added component update timeout to Playwright user story test to prevent eaten input. aba4b28
- [x] Updated variable names based on latest owner feedback.

Thank you @wei for your tremendous support; I look forward to replying to your review. 

Edit: additional context

### CODE REVIEW

**Comment by KemingHe** [comment]
Ref: step 1/2 to resolve #373 .

File: src/hooks/useRouteResources.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+// biome-ignore format: added alignment for clarity.
+export interface RouteResources {
+  clientRouter      : ReturnType<typeof useRouter>
+  repoOwner         : string
+  repoName          : string
+  currentPath       : string
+  searchParamsString: string
+}
+
+export function UseRouteResources(): RouteResources {
```
Comment (wei) [inline src/hooks/useRouteResources.ts:17]: Let's rename to useRouteResources to be consistent with react hook naming convention

File: src/components/configuration/config.tsx [hunk 2]
```diff
@@ -58,76 +68,71 @@ const Config = ({ repository }: ConfigProp) => {
...
         }
-      }
+      })
+      setConfig({ ...config, ...newConfig })
     }
 
-    router.events.on('routeChangeComplete', handleRouteChange)
-    handleRouteChange(router.asPath)
+    handleRouteChange(searchParamsString)
+  }, [searchParamsString])
```
Comment (wei) [inline src/components/configuration/config.tsx:132]: Excellent refactor!

File: app/layout.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  // userScalable: false,
+}
+
+export default function RootLayout({
+  children,
+}: Readonly<{ children: ReactNode }>): JSX.Element {
+  return (
+    <html lang="en">
+      <head>
+        <meta name="viewport" content="minimal-ui" />
```
Comment (wei) [inline app/layout.tsx:47]: does the export viewport overwrite this meta tag?

File: app/api/image/route.ts
Comment (wei) [inline app/api/image/route.ts:1]: I believe we are expected to always returning svg for this endpoint, please correct me if I'm wrong.

**Review by wei — COMMENTED** [review]
This is wonderful 💯! I see so many great improvements. 

I left a few questions above please review and lmk your thoughts. 

I'll conduct a more thorough review when I'm on my computer tomorrow. 🙌

File: app/layout.tsx (hunk 3 above)
Comment (KemingHe) [reply app/layout.tsx:47]: Actually no, the NextJS `Viewport` and `Metadata` export merely **appends** the corresponding elements to the `<head>`. Can confirm my inspecting my browser.

File: src/hooks/useRouteResources.ts (hunk 1 above)
Comment (KemingHe) [reply src/hooks/useRouteResources.ts]: Resolved in latest commit. Thank you for point it out.

File: app/layout.tsx (hunk 3 above)
Comment (wei) [reply app/layout.tsx:47]: Looks like we can remove minimal-ui as it’s been deprecated a while~

File: package.json [hunk 4]
```diff
@@ -30,21 +30,20 @@
...
     "clsx": "^2.1.1",
     "copee": "^1.0.6",
     "hero-patterns": "^2.1.0",
     "is-ci": "^4.1.0",
-    "next": "^14.2.8",
-    "react": "^18.3.1",
-    "react-dom": "^18.3.1",
+    "next": "15.1.0",
+    "react": "19.0.0",
+    "react-dom": "19.0.0",
```
Comment (wei) [inline package.json:42]: Should we prepend with `^` to match future versions in the major release?

File: package.json [hunk 5]
```diff
@@ -82,5 +81,10 @@
       "last 1 safari version"
     ]
   },
-  "packageManager": "pnpm@9.15.0+sha512.76e2379760a4328ec4415815bcd6628dee727af3779aaa4c914e3944156c4299921a89f976381ee107d41f12cfa4b66681ca9c718f0668fa0831ed4c6d8ba56c"
+  "packageManager": "pnpm@9.15.0+sha512.76e2379760a4328ec4415815bcd6628dee727af3779aaa4c914e3944156c4299921a89f976381ee107d41f12cfa4b66681ca9c718f0668fa0831ed4c6d8ba56c",
+  "pnpm": {
+    "overrides": {
+      "@types/react": "19.0.1"
+    }
+  }
```
Comment (wei) [inline package.json:89]: What's this override for?

File: tsconfig.json
Comment (wei) [inline tsconfig.json:1]: Did something in biome change to trigger this update?

File: package.json (hunk 4 above)
Comment (KemingHe) [reply package.json:42]: I agree, adding in next commit.

File: package.json (hunk 5 above)
Comment (KemingHe) [reply package.json:89]: This is the raw result from the codemod. I also agree that this is odd. Currently testing what happens if I remove it.

Comment (KemingHe) [reply package.json]: All tests (unit and behavior) passing after removal of the override part. Determined safe removal for the time being.

No mentioning of this change found on https://nextjs.org/blob/next-15 nor the official upgrade guide.

File: tsconfig.json
Comment (KemingHe) [reply tsconfig.json:1]: Biome is not part of the update codemod. Can confirm biome also did not edit that portion of `package.json` as re-running the linter/formatter after deletion does not get it back.

**Comment by KemingHe** [comment]
`server-only` is a react (not next) marker that enforces usage of the module that imports it to run on server-side. It's widely used to explicitly protect code involving server-side secrets.

I'll let you ultimately decide whether to keep it. It's extremely light-weight w/ no versioning/upgrade needed.

**Comment by wei** [comment]
👍 server-only is definitely a good DX and build time improvement. https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#keeping-server-only-code-out-of-the-client-environment

File: app/api/graphql/route.ts [hunk 6]
```diff
@@ -1,25 +1,21 @@
...
+  const response = await fetch(GITHUB_GRAPHQL_ENDPOINT, {
     method: 'POST',
     headers: {
       Accept: 'application/json',
       Authorization: `bearer ${process.env.GITHUB_TOKEN}`,
       'content-type': 'application/json',
     },
     body: req.body,
+    // @ts-expect-error: 'duplex' is not part of the RequestInit type but required by GraphQL.
+    duplex: 'half',
```
Comment (wei) [inline app/api/graphql/route.ts:18]: Do you have any docs on why this is now needed?

**Review by wei — CHANGES_REQUESTED** [review]
Could you apply these two changes for Next15 upgrade?
* https://nextjs.org/docs/app/api-reference/file-conventions/metadata/manifest
* https://nextjs.org/docs/app/api-reference/file-conventions/metadata/app-icons

File: app/api/graphql/route.ts (hunk 6 above)
Comment (KemingHe) [reply app/api/graphql/route.ts:18]: Yes, this is actually an official nodejs fetch api "issue": https://github.com/nodejs/node/issues/46221

(Also encountered by graphql-yoga folks: https://github.com/dotansimha/graphql-yoga/discussions/3175)

More straightforwardly, you can try to remove it and read the error message about needing it. I personally am not too well-versed on the root cause yet, even after researching the error.

Comment (wei) [reply app/api/graphql/route.ts:18]: Thanks for the link! Did this only become a problem after migrating to app router?

Comment (KemingHe) [reply app/api/graphql/route.ts:18]: It appears so @wei , my speculation is that it's caused by the now explicit use of React server component as a POST endpoint. In app router, we didn't even enforce 'server-only'. 

Another possibility is by some inner working of the app router, not directly related to React. Both (and maybe more) are just my best guesses.

**Comment by KemingHe** [comment]
Hi @wei , I used [`https://realfavicongenerator.net/`](https://realfavicongenerator.net/) to double-confirm my favicon and manifest updates are good. I did **not** re-generate/format any of the files, only copy/rename.

P.S. It's a known issue that their site's `favicon.ico` check is bugged recently, everything else works fine, so no worries.

**Comment by KemingHe** [comment]
I also found this unexpected behavior where the import breaks for the `wasm` module as soon as I touched the `public` dir. The fix is to revert (only) them to relative imports. 

I suspect this is related to how next caches/imports? `public` dir and how the behavior changed once `favicon.ico` and `manifest.json` got moved out.

**Comment by wei** [comment]
@KemingHe  That all sounds and looks great! I was previewing the feature branch and noticed the selected font is not loaded for the preview. Everything else is working well. Could you help confirm the font is loaded correctly?

image

**Comment by KemingHe** [comment]
Hi @wei , assuming you are referring to the latest [`https://deploy-preview-450--github-socialify.netlify.app/`](https://deploy-preview-450--github-socialify.netlify.app/) preview, here's what I found:

- All image urls are working (with custom fonts).
- Only the Inter/Non-inter font change is working on preview config page (page refresh doesn't fix it).

I have a few ideas on the root cause, but before we get into that, what if you try to create a branch off the current working `master` and deploy a preview without any changes to see if the issues are also present and consistent?

If so I argue that we are actually good to merge to `master`. Otherwise we can keep investigating. What do you think?

**Comment by wei** [comment]
@KemingHe good point!

I just checked the master branch deployment, permalink: https://67611dae31fa580008b7c232--github-socialify.netlify.app and confirmed fonts are working correctly.

**Comment by wei** [comment]
I believe the permalink act the same as the pull request preview deployments.

---

Image APIs add fonts within the svg so it's slightly different:

https://github.com/wei/socialify/blob/811ccd2d21b2df3697755326a0ca82c207c84b87/common/renderSVG.tsx#L25

Whereas the preview page loads font like this:

https://github.com/wei/socialify/blob/811ccd2d21b2df3697755326a0ca82c207c84b87/src/components/preview/preview.tsx#L129-L140

I noticed in master, these two font stylesheet links are added to head, but not in the pull request preview domain:
image

I see no changes around these lines here so it might be related to how <Head> and link headers are handled.

RE: font fallback is not used as defined in root layout as the font family was overwritten without a fallback:
image

**Comment by KemingHe** [comment]
**YES** using vanilla `<head>` solves the issue, even in preview. Committing the fix right now.

**Comment by wei** [comment]
We can't have two `<head>`s in an html, the nextjs `<Head>` is a helper component acting a portal to move whatever is in there to the actual html's `<head>`.

**Comment by KemingHe** [comment]
I propose we (I) modify the current Playwright `simpleUserStory.spect.ts` to cover this case. This will require a snapshot update. What do you think?

**Comment by wei** [comment]
Confirmed `next/head` does not work with app router per https://github.com/vercel/next.js/discussions/53529#discussioncomment-6628991

I'm working on an alternative.

**Comment by KemingHe** [comment]
@wei , my latest commit is a (like you said) "bad practice" fix, as it uses the html `<head>` in `preview.tsx`. Still, please use the updated user story snapshots to confirm your better solution. Excited for the great news.

**Review by wei — APPROVED** [review]
@KemingHe ✅ Approved!! I'm so excited.

**Pull request merged**